### PR TITLE
Leaves - Dominique

### PR DESCRIPTION
**Pull request by dtaylor73**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method called "self.sort_by_category?(category)" that sorted that works in each category according to how many votes it received. 
Describe how you approached testing that model method. What edge cases did you come up with? | I tested to see if there is only 1 work, that it would return only 1 work. Since there could only be one work, I wanted see is if the method would still execute.
What are `session` and `flash`? What is the difference between them? | Session is a special type of hash-like-object that we utilize to keep track of data throughout a users "session", which normally ends when they close their browser. Flash only last for one request cycle. Flash is also a special type of hash. 
What was one thing that you gained more clarity on through this assignment? | How to use partials, strong params, controller filters and how to test model methods. I learned a lot!
What is the Heroku URL of your deployed application? | https://media-ranker-app.herokuapp.com/ 
Do you have any recommendations on how we could improve this project for the next cohort? | It was harder than I thought it was going to be

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+  resources :works
+  resources :users, only: [:index, :show]
+
+  post "/works/:id/upvote", to: "works#upvote", as: "upvote"
+
+
+  get "/login", to: "users#login_form", as: "login_form"
+  post "/login", to: "users#login"
+  post "/logout", to: "users#logout", as: "logout"
```
Comment (dHelmgren) [inline config/routes.rb]: Love the restraint here! Just the routes you need!

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  </thead>
+  <tbody>
+  <%@albums.each do |album|%>
+    <tr>
+      <td><%=album.votes.count%></td>
+      <td><%= link_to album.title, work_path(album)%></td>
+      <td><%=album.creator%></td>
+      <td><%=album.publication_year%></td>
+      <td><%=link_to "upvote", upvote_path(album), :method => :post, :class => "btn btn-primary"%></td>
+    </tr>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:48]: This code gets repeated. Is there a way you could use a view partial to Dry this up?

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
+class WorksController < ApplicationController
+
+  # Order matters here
+  before_action :find_work, only: [:show, :edit, :update, :destroy]
+  before_action :if_work_missing, only: [:show, :edit, :destroy]
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:5]: Nice!

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return 
+  end
+
+  def upvote
+
+    if session[:id].nil?
+      flash[:error] = "A problem occured: You must be logged in to do that."
+      redirect_to works_path
+      return 
+    end
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:63]: Seems like a good candidate for a controller filter, no?

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    end
+  end 
+
+  it "returns 1 work if there is only 1 work" do
+    result = Work.find_top_10("album")
+
+    expect(result.length).must_equal 1
+  end
+
+  it "returns a list of works even if none of them have votes" do
```
Comment (dHelmgren) [inline test/models/work_test.rb:125]: YEAH!!!!!

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes, and so nice too!
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes, but somehow old accounts have more than 1 vote per work?
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | no, see note!
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Great job over all! this is a very complete and comprehensive project! Take a look at my notes to see what can be improved, and keep up the hard work!
